File: app/functions/src/dto/tags.ts
export const tagSchemaKey = "tags";

export type TagJSON = {
  key?: string;
  fallback?: string;
  promoted?: boolean;
  popularity?: number;
  localizations?: TagLocalizationJSON[];
  topic?: TagTopicJSON;
};

export type TagTopicJSON = {
  fallback?: string;
  localizations?: TagLocalizationJSON[];
  isEnabled?: boolean;
};

export type TagLocalizationJSON = {
  locale?: string;
  value?: string;
};

/**
 * Represents a tag.
 * @class
 * @property {string} key the tag key.
 * @property {string} fallback the fallback value.
 * @property {boolean} promoted true if the tag is promoted.
 * @property {TagLocalization[]} localizations the localizations.
 * @returns {Tag} the tag.
 */
export class Tag {
  key: string;
  fallback: string;
  promoted: boolean;
  popularity: number;
  localizations: TagLocalization[];
  topic?: TagTopic;

  /**
   * Creates a new tag.
   * @param {TagJSON} json the json.
   * @returns {Tag} the tag.
   * @constructor
   */
  constructor(json: TagJSON) {
    this.key = json.key || "";
    this.fallback = json.fallback || "";
    this.promoted = json.promoted || false;
    this.popularity = json.popularity || -1;
    this.localizations = json.localizations?.map((e) => new TagLocalization(e)) || [];
    this.topic = json.topic ? new TagTopic(json.topic) : undefined;
  }

  /**
   * Gets a list of tags from a json array.
   * @param {TagJSON[]} data the json array.
   * @returns {Tag[]} the tags.
   */
  static fromJsonArray(data: any[]): Tag[] {
    return data.map((e) => new Tag(e));
  }

  /**
   * Gets a tag from json.
   * @param {TagJSON} json the json.
   * @returns {Tag} the tag.
   */
  static fromJSON(json: TagJSON): Tag {
    return new Tag(json);
  }
}

/**
 * Represents a tag localization.
 * @class
 * @property {string} locale the locale.
 * @property {string} value the value.
 */
export class TagLocalization {
  locale: string;
  value: string;

  /**
   * Creates a new tag localization.
   * @param {TagLocalizationJSON} json the json.
   * @returns {TagLocalization} the tag localization.
   */
  constructor(json: TagLocalizationJSON) {
    this.locale = json.locale || "";
    this.value = json.value || "";
  }
}

export class TagTopic {
  fallback: string;
  localizations: TagLocalization[];
  isEnabled: boolean;

  constructor(json: TagTopicJSON) {
    this.fallback = json.fallback || "";
    this.localizations = json.localizations && Array.isArray(json.localizations) ? json.localizations?.map((e) => new TagLocalization(e)) : [];
    this.isEnabled = json.isEnabled || false;
  }
}

File: app/lib/dtos/database/activities/tags.dart
// Dart imports:
import 'dart:convert';

// Flutter imports:
import 'package:flutter/material.dart';

// Package imports:
import 'package:freezed_annotation/freezed_annotation.dart';

// Project imports:
import '../common/fl_meta.dart';

part 'tags.freezed.dart';
part 'tags.g.dart';

@freezed
class Tag with _$Tag {
  const factory Tag({
    @JsonKey(name: '_fl_meta_') FlMeta? flMeta,
    @Default('') String key,
    @Default('') String fallback,
    @Default(-1) int popularity,
    @Default(false) bool promoted,
    @Default([]) @JsonKey(fromJson: TagLocalization.fromJsonLocalizations, toJson: TagLocalization.toJsonLocalizations) List<TagLocalization> localizations,
    TagTopic? topic,
  }) = _Tag;

  factory Tag.fromJson(Map<String, dynamic> json) => _$TagFromJson(json);
  static List<Tag> fromJsonList(List<dynamic> json) => json.map((e) => Tag.fromJson(e)).toList();
}

@freezed
class TagLocalization with _$TagLocalization {
  const factory TagLocalization({
    @Default('') String locale,
    @Default('') String value,
  }) = _TagLocalization;

  static List<TagLocalization> fromJsonLocalizations(dynamic json) {
    if (json is! Iterable || json.isEmpty) {
      return [];
    }

    return json.map((e) => TagLocalization.fromJson(e)).toList();
  }

  static String toJsonLocalizations(List<TagLocalization> localizations) {
    return jsonEncode(localizations.map((e) => e.toJson()).toList());
  }

  static TagLocalization fromJsonSafe(Map json) {
    if (json.isEmpty) {
      return const TagLocalization();
    }

    return TagLocalization.fromJson(json as Map<String, dynamic>);
  }

  factory TagLocalization.fromJson(Map<String, dynamic> json) => _$TagLocalizationFromJson(json);
}

@freezed
class TagTopic with _$TagTopic {
  const factory TagTopic({
    @Default('') String fallback,
    @Default([]) @JsonKey(fromJson: TagLocalization.fromJsonLocalizations, toJson: TagLocalization.toJsonLocalizations) List<TagLocalization> localizations,
    @Default(false) bool isEnabled,
  }) = _TagTopic;

  factory TagTopic.fromJson(Map<String, dynamic> json) => _$TagTopicFromJson(json);
  factory TagTopic.empty() => const TagTopic();

  static List<TagTopic> fromJsonList(dynamic json) {
    if (json is! Iterable || json.isEmpty) {
      return [];
    }

    return json.map((e) => TagTopic.fromJson(e)).toList();
  }
}

class TagHelpers {
  static const String _kPromotedKey = 'promoted';
  static const String _kFallbackKey = 'fallback';
  static const String _kKeyKey = 'key';
  // helper to create the tags that should be appended when an activity is promoted
  static List<String> createPromotedTags({required String userId}) => [createPromotedTag(), createPromotedTag(userId: userId)];

  /// helper to create the tag to be in the list of tags to show an activity is promoted
  static String createPromotedTag({String? userId}) => '$_kPromotedKey${userId == null ? '' : '-$userId'}';

  /// helper to determine if a tag (as a string) represents an activity that is promoted
  static bool isPromoted(String tag) => tag.startsWith(_kPromotedKey);

  static String getTagLocalizedName(Tag tag, Locale locale) {
    if (tag.localizations.any((TagLocalization localization) => localization.locale == locale.languageCode)) {
      return tag.localizations.firstWhere((TagLocalization localization) => localization.locale == locale.languageCode).value;
    }

    return tag.fallback;
  }

  static bool matches(Tag tag, Map<String, dynamic> map) {
    if (tag.key != map[_kKeyKey]) {
      return false;
    }

    if (tag.fallback != map[_kFallbackKey]) {
      return false;
    }

    if (tag.promoted != map[_kPromotedKey]) {
      return false;
    }

    // Check if all of the localizations match
    for (var localization in tag.localizations) {
      if (!map.containsKey(localization.locale) || map[localization.locale] != localization.value) {
        return false;
      }
    }

    return true;
  }
}
